sentry-core/src/session.rs

```diff
@@ -0,0 +1,316 @@
+//! Release Health Sessions
+//!
+//! https://develop.sentry.dev/sdk/sessions/
+
+use std::borrow::Cow;
+use std::sync::Arc;
+use std::time::Instant;
+
+use crate::protocol::{Event, Level, User};
+use crate::scope::StackLayer;
+use crate::types::{DateTime, Utc, Uuid};
+
+/// Represents the status of a session.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum SessionStatus {
+    Ok,
+    Crashed,
+    #[allow(dead_code)]
+    Abnormal,
+    Exited,
+}
+
+pub enum SessionUpdate {
+    NeedsFlushing(Session),
+    Unchanged,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Session {
+    session_id: Uuid,
+    status: SessionStatus,
+    errors: usize,
+    user: Option<Arc<User>>,
+    release: Cow<'static, str>,
+    environment: Option<Cow<'static, str>>,
+    started: Instant,
+    started_utc: DateTime<Utc>,
+    duration: Option<f64>,
+    init: bool,
+    dirty: bool,
+}
+
+impl Session {
+    pub fn from_stack(stack: &StackLayer) -> Option<Self> {
+        let options = stack.client.as_ref()?.options();
+        Some(Self {
+            session_id: Uuid::new_v4(),
+            status: SessionStatus::Ok,
+            errors: 0,
+            user: stack.scope.user.clone(),
+            release: options.release.clone()?,
+            environment: options.environment.clone(),
+            started: Instant::now(),
+            started_utc: Utc::now(),
+            duration: None,
+            init: true,
+            dirty: true,
+        })
+    }
+
+    pub(crate) fn update_from_event(&mut self, event: &Event<'static>) -> SessionUpdate {
+        if self.status != SessionStatus::Ok {
+            // a session that has already transitioned to a "terminal" state
+            // should not receive any more updates
+            return SessionUpdate::Unchanged;
+        }
+        let mut has_error = event.level >= Level::Error;
+        let mut is_crash = false;
+        for exc in &event.exception.values {
+            has_error = true;
+            if let Some(mechanism) = &exc.mechanism {
+                if let Some(false) = mechanism.handled {
+                    is_crash = true;
+                    break;
+                }
+            }
+        }
+
+        if is_crash {
+            self.status = SessionStatus::Crashed;
```

Comment:
If this is different from a previous status should it not be considered dirty?

Reply:
it might not be totally obvious, but `is_crash` implies `has_error`, which makes it `dirty`.